### DISCUSSION TITLE
What is the best way to create animated jpegxl image?

### DISCUSSION THREAD

**Discussion by jetgogol** [libjxl/libjxl#3495] [Q&A]
As table on jpegxl.info says jxl should have animation feature, but I cant figure out how to do it in best way by myself. I tried this:

    $ convert file.gif file.jxl
    online at https://mconverter.eu/convert/gif/jxl/
    $ ffmpeg -i file.gif file.jxl
    cjxl file.gif file.jxl

But all this ways make static image in mpv and in librewolf (built-in jpeg xl). As I see in jxlinfo images have there. May be jxllib now not included for support animation?

If I done something wrong, I want to hear your version.

**Comment by mo271** [comment]
There are some ways of creating animated jxl files, for instance
```
$ wget https://upload.wikimedia.org/wikipedia/commons/2/2c/Rotating_earth_%28large%29.gif
$ cjxl Rotating_earth_\(large\).gif earth.jxl
$ jxlinfo earth.jxl 
JPEG XL animation, 400x400, (possibly) lossless, 8-bit RGB+Alpha
Color space: RGB, D65, sRGB primaries, sRGB transfer function, rendering intent: Perceptual
frame: full image size, duration: 90.0 ms
frame: 280x281 at position (59,31), duration: 90.0 ms
[...omitted some lines...]
frame: 278x281 at position (60,31), duration: 90.0 ms
Animation length: 3.960 seconds (looping)
```
There might be some viewers, which don't fully supported viewing animated jxl files yet. In order to check if you created a jxl animation, you could use `jxlinfo` or decode the jxl animated image to an image format that supports animation, e.g. apng, 
```
$ djxl earth.jxl earth.apng
```
And then use an image viewer that supports animated pngs.

**Comment by jetgogol** [comment]
Thank you for answer! Do you have any list of programs that now supports jxl animations? I want to use it on my website, but may be I should wait bit of time.

**Comment by perk11** [comment]
The only way I found to convert a video or series of images into animated jxl currently was to use `ffmpeg` to create `apng` first and then using `cjxl` to convert it to `jxl`. The `gif` format is limited to 256 colors.

Also `gwenview` will play animated jxl files.

**Reply by dante-e-v** [reply]
> The only way I found to convert a video or series of images into animated jxl currently was to use `ffmpeg` to create `apng` first and then using `cjxl` to convert it to `jxl`. The `gif` format is limited to 256 colors.
> 
> Also `gwenview` will play animated jxl files.

I guess we stumbled on the same solution. Though now `cjxl` seems a lot better at converting animated image formats such as a GIF or APNG. I doubt it will work on an AVIF though since that uses a video codec which `ffmpeg` should be handling but they still haven't gotten around to converting anything to JXL. That does mean you have to do the workaround still. 

My problem is that converting from APNG to JXL ruins the timing and the delays between each image is like playing things in slow motion. I was hoping they would they would have fixed that by now since converting from GIF to JXL seems to work flawlessly now. Now perhaps ImageMagick can help with creating the proper image sequence delays so I'm going to try testing that soon when I have the time.

There's also the issue that `cjxl` still doesn't even try to convert an animated WEBP image whether static or animated and instead you get this:
```
JPEG XL encoder v0.11.1 794a5dcf [AVX2,SSE4,SSE2]
Getting pixel data failed.
```
At least there's the workaround of just using ImageMagick to convert a WEBP static into a PNG and then use `cjxl` to finish up since ImageMagick can't take any options for JXL such as distance or effort and I did try using the `--quality` flag but it just threw an error. Though if you are fine with a lower quality image you can have ImageMagick convert WEBP to JXL or even with `ffmpeg` but I don't think it can take JXL option flags either.

Maybe we'll see things develop faster and actually work properly since Google has finally admitted defeat and is going to start putting effort into supporting JXL after the PDF group announced they are adding JXL to the PDF format.

**Reply by dante-e-v** [reply]
If I were to pick a winner right now for animated image formats I would pick AVIF. JXL for static for sure but you just can't beat the efficiency of AVIF using a video codec and I doubt JXL is planning to create its own. If anything they would just use the same AV1 codec since that's the latest and greatest open video codec standard.

But I mean if you do like having things in an old fashioned image sequence I would certainly go with JXL. It's half the size or more of GIFs or APNGs. But again there's no way to actually edit that sequence right now with the tools as they are.

**Reply by Traneptora** [reply]
As of now, you can do `ffmpeg -i input -c libjxl_anim -f image2pipe animated.jxl`
